examples/post-sync/main.go

```diff
@@ -0,0 +1,352 @@
+
+		if isErrorCondition {
+			errorMessages = append(errorMessages,
+				fmt.Sprintf("%s: %s", conditionType, condition.Message))
+			if commit == "" && condition.Commit != "" {
+				commit = condition.Commit
+			}
+		}
+	}
+
+	return strings.Join(errorMessages, "\n"), commit
+}
+
+// isSuccessfulSync checks if the rsync status no errors in rsync status and all commits match.
+func isSuccessfulSync(status v1beta1.Status) bool {
+	allCommitsMatch := status.Source.Commit != "" &&
+		status.Source.Commit == status.Rendering.Commit &&
+		status.Rendering.Commit == status.Sync.Commit
+
+	return allCommitsMatch && !hasStatusError(status)
+}
+
+// hasStatusError checks if there are any errors in the rsync status.
+func hasStatusError(status v1beta1.Status) bool {
+	return status.Source.Errors != nil ||
+		status.Rendering.Errors != nil ||
+		status.Sync.Errors != nil
+}
+
+// extractErrorAndCommitFromStatus extracts errors and the associated commit from the status.
+func extractErrorAndCommitFromStatus(status v1beta1.Status) (string, string) {
+	if status.Source.Errors != nil {
+		return aggregateErrors(status.Source.Errors), status.Source.Commit
+	}
+	if status.Rendering.Errors != nil {
+		return aggregateErrors(status.Rendering.Errors), status.Rendering.Commit
+	}
+	return aggregateErrors(status.Sync.Errors), status.Sync.Commit
+}
+
+// aggregateErrors aggregates error messages from a slice of ConfigSyncError
+// and returns a single error message.
+func aggregateErrors(errors []v1beta1.ConfigSyncError) string {
+	if len(errors) == 0 {
+		return ""
+	}
+
+	var errorMessages []string
+	for _, err := range errors {
+		errorMessages = append(errorMessages, fmt.Sprintf("Code: %s, Message: %s", err.Code, err.ErrorMessage))
+	}
+
+	return "aggregated errors: " + strings.Join(errorMessages, "\n")
+}
+
+// Update the StatusTracker to use SyncID
+type StatusTracker struct {
+	mu   sync.Mutex
+	seen map[SyncID]map[string]struct{}
+}
+
+// NewStatusTracker creates a new StatusTracker instance.
+func NewStatusTracker() *StatusTracker {
+	return &StatusTracker{
+		seen: make(map[SyncID]map[string]struct{}),
+	}
+}
+
+// compositeKey generates a unique key for a commit and message.
+// Using a simple concatenation with a separator. Hashing could be used for long messages.
+func compositeKey(commit, message string) string {
+	return commit + "::" + message
+}
+
+// Update the IsLogged and MarkLogged methods
+func (s *StatusTracker) IsLogged(syncID SyncID, commit, message string) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := compositeKey(commit, message)
+	if commitMap, ok := s.seen[syncID]; ok {
+		_, logged := commitMap[key]
+		return logged
+	}
+	return false
+}
+
+func (s *StatusTracker) MarkLogged(syncID SyncID, commit, message string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := compositeKey(commit, message)
+	if _, ok := s.seen[syncID]; !ok {
+		s.seen[syncID] = make(map[string]struct{})
+	}
+	s.seen[syncID][key] = struct{}{}
+}
+
+func main() {
```

Comment:
Recommend having a run func that returns an error, and then just have the main func handle converting the error to an exit code and error log message.

Reply:
This also makes it easier to unit test the run func

Reply:
Splitted